Implement range ignore (Fixes #5287)

chaitan94 opened the pull request:
## Summary

JavaScript code within `// prettier-ignore-start` and `// prettier-ignore-end` will be excluded in the abstract syntax tree from formatting.

For example:

<!-- prettier-ignore -->
```js
// prettier-ignore-start
let i_like         = 1;
let my_assignments = 2;
let sorted         = 3;
// prettier-ignore-end
let but_not    = 4;
let after      = 5;
let ignore_end = 6;
```

will be transformed to:

```js
// prettier-ignore-start
let i_like         = 1;
let my_assignments = 2;
let sorted         = 3;
// prettier-ignore-end
let but_not = 4;
let after = 5;
let ignore_end = 6;
```

---------

- [x] I’ve added tests to confirm my change works.
- [x] I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

fisker commented on changelog_unreleased/javascript/pr-8926.md:
```diff
@@ -1,4 +1,4 @@
-#### Support range ignore for javascript code
+#### Implement range ignore (javascript) ([#8926](https://github.com/prettier/prettier/pull/8926) by [@chaitan94](https://github.com/chaitan94))
```
```suggestion
#### Implement range ignore ([#8926](https://github.com/prettier/prettier/pull/8926) by [@chaitan94](https://github.com/chaitan94))
```

fisker replied:
The changelog already categorized

chaitan94 replied:
Makes sense, I removed the javascript part.

fisker commented:
**Prettier pr-8926**
[Playground link](https://deploy-preview-8926--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAzArlMMCW0AEqEEAFMADpQCG+AvPgAwA0+lA9G-gA4BOcMuODwC0OAOZQIfYQGcYVHjEoAjfGvUb6ARnade-QSPGTpCACaUNVjQF8AlPgpQbIJiAhdc0GclAKeEADuAAoKCD4oVAA2gVQAnj5uyjxUYADW-ADKXKk4UGLIMDzocG5wALbKcGZm1QAyVPnoVGJwAGJS5VQCeQWR6DAQriAAFjDlUQDqIzjwMjlgcJnhszgAbrNxyOAyiSB5MkIwwSliXcio0YduAFYyAB4AQinpWVTlcHV5cBdXpSB3e6ZXpROAARXQEHgvyi1xAOR4hx422UVCqUWGvDyMEmODMMBGyAAHMx4QFDpMUlxtvokWsfm4AI6Q+AnDwREBUGTCKBwarVYZ8Zk4PgnFrnJCXWH-Q7lHCFYoykHglk-SV-NzyZS4-GEpAAJk1KRwUV6AGEIOUJSA4DIAKzDdCHAAqaIiUrhaxKAEkoLVYJkwDwcJ4AIJ+zIwOKgmGHGw2IA)
```sh
--parser babel
```

**Input:**
```jsx
function foo({
a = 0, 
// prettier-ignore-start
b          = 1
// prettier-ignore-end
             }) {
}
```

**Output:**
```jsx
TypeError: Cannot read property 'declaration' of undefined
    at $t (https://deploy-preview-8926--prettier.netlify.app/lib/parser-babel.js:1:43406)
    at https://deploy-preview-8926--prettier.netlify.app/lib/standalone.js:15:107470
    at ri (https://deploy-preview-8926--prettier.netlify.app/lib/standalone.js:15:107552)
    at https://deploy-preview-8926--prettier.netlify.app/lib/standalone.js:15:107033
    at Object.printComments (https://deploy-preview-8926--prettier.netlify.app/lib/standalone.js:15:103467)
    at e (https://deploy-preview-8926--prettier.netlify.app/lib/standalone.js:15:107014)
    at Ga.call (https://deploy-preview-8926--prettier.netlify.app/lib/standalone.js:15:105253)
    at Uf (https://deploy-preview-8926--prettier.netlify.app/lib/standalone.js:15:210708)
    at $f (https://deploy-preview-8926--prettier.netlify.app/lib/standalone.js:15:210613)
    at https://deploy-preview-8926--prettier.netlify.app/lib/standalone.js:15:222883
```

chaitan94 commented:
Thanks for the quick review @fisker. I've also noticed that the example given in #7737 also doesn't work properly.

[Playground link](https://deploy-preview-8926--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEB6VACADgJzjGASzhwFpCBzKCPUgZxgEMcYAdKAG3w0IH0PCAazgZRYsQF4MARgDc7LjAwBbAJ69GdOpSjKEMOhikAmeZ250a8ACbi7UgMyyMAMwCuUMEWgYAHgAoASnFgAF92dGw8AmIyHRo4UgRrBW4AIzcYXmolSQwAFjNFDEYXeBw7IwwAViLueLxeZKqANmd3T28oPyCQ8KgIzFx8IhJyKgT6JhZ2RiwsAEF5gTBGLoA6LAgGf3ZRAAMAEmAFgAUASQAlCEy4ABE4F0IoQi66TbWACwAxGmvb-wYM7nB5PF5vdY4G7wd4UfCnZLPCgAYU+jCgcLovxwABVPng6J8IBxrIFQvsADR7DD+PAARwpGAJwQkAD4MMBqaIAG6MATWNZwS5wOluOA7emMgmM4Gg56vQjQd5Q26w+GIjGo9GY7F4glEklSkXrNIQayqQJmUT9S2zeZLLArNaKqCbbYwXbdDBHE4Xf7wOXgl3vLBfbH+uCA2WPeUQlUw9ZwmAIqDWJFajHi3X48UG0nkqle2kio10Fnszlenl8wgC+DC0Xij2SpnimUXQMKpWQ6HixPq1PptGZrE0PW54nWI10k1mi1WjA2syRYYxMYNRLJdggCkgCBYN7IUDMKEAd1OzAQdGQID5p8Yqmvu7SOEYYGEMAAyqGwEjkDAcDFXc4GUNI4GsaxwIAGW1NxGDhbFlDWIgMRvRhMggHcQE+GBlA4AB1T5XnFH84E-K8FW5V5VBvMAtCw546BIZNXwoJDkBcPkmN3AArOhfAAIVfd98E-Rg9Cg544A4ri4F4-jPyRLgAEU3AgeAZI4biQFDHAmJwG80kYMCOCw3BnhgfDaxgT5kAADgABl3XBtjgfDXywG9hn07lpN3UV1LgU4oSwa8UE0UgoDgcDwKw+k3EIPBgvg9ikE4rS5JAJjlEIf9AMy7QMRUtSNLS2TdyYNIrOsGzkGMCrX0IARNQgZRUpAcVqiwtwmJxYywvS7TuTFc5U30T8wBwQgDwWVNPxgVQuE0pjQlCIA)

### Input
```js
// prettier-ignore-start
appApplication.post(
  `${APIRouteDefinitions.pathForRoute( APIDefinitions.routes.getPendingChangesForThreshold)}`,
  (req, res) => {
    validateRequest(req, res, APIDefinitions.routes.getPendingChangesForThreshold, req.body);
  }
);
appApplication.post(
  `${APIRouteDefinitions.pathForRoute( APIDefinitions.routes.getPendingChangesForThreshold)}`,
  (req, res) => {
    validateRequest(req, res, APIDefinitions.routes.getPendingChangesForThreshold, req.body);
  }
);
// prettier-ignore-end
```
### Output
```js
// prettier-ignore-start
appApplication.post(
  `${APIRouteDefinitions.pathForRoute( APIDefinitions.routes.getPendingChangesForThreshold)}`,
  (req, res) => {
    validateRequest(req, res, APIDefinitions.routes.getPendingChangesForThreshold, req.body);
  }
);
appApplication.post(
  `${APIRouteDefinitions.pathForRoute(
    APIDefinitions.routes.getPendingChangesForThreshold
  )}`,
  (req, res) => {
    validateRequest(
      req,
      res,
      APIDefinitions.routes.getPendingChangesForThreshold,
      req.body
    );
  }
);
// prettier-ignore-end
```
### Expected
Same as input. That is, both groups should be ignored.

I will try and find out what is going wrong. Do you have any suggestions as to what exactly I should be looking at?

Once these issues are fixed, I will be including these test case as well in the PR.

fisker commented:
`options[Symbol.for('comments')]` have all comments, you might need it.

chaitan94 commented:
@fisker I added more tests, changelog title, and fixed the code to work correctly with the example I mentioned.

I think the implementation is still rough around more complicated use cases, but seems to work on basic usages which devs might use this feature for. i.e, for ignore big chunks of code.

Let me know if anything else can be improved.

chaitan94 commented:
@evilebottnawi I have added more tests for javascript, accordingly to the scenarios you mentioned. Additionally I added one basic test for typescript as well.

chaitan94 commented:
@evilebottnawi @fisker is there anything else I could do?

fisker commented:
**Prettier pr-8926**
[Playground link](https://deploy-preview-8926--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEBDABAXnQHSu9YAeiPQAcAnOGGASzgoFpaBzKCKxgZxlQpjwFUAGjwlyVGvSat2nBABNBBFarXqNmgL4BuPHgzZxlanQbM2HON179lwZSLGkTU87KuNFyzb7-adEGEQCDI6aC5kUD4KCAB3AAU+BEiUVAAbONQAT0jggCMKVDAAa2oAZTJi2igWZBgKAFc4YLgAW3y4BQUugBlUWsbUFjgAMQ421Cla5DRGmAggkAALGDb0gHVl2nguKrA4cpSd2gA3HezZsC48kBquBhgEopZJ5AAzDIfggCsuAA8AEJFUoVVBtOC9GpwD5fFogP7-co1FjpOAARUaEHgsPS3xAVQoDwos3yqE66SWlBqMA2tAUMGWyAAHAAGYKUCAPDZFMizEzE04w4IARyx8GeoVSaC4jCgcC6XSWVDFtCoz2GbyQnzx8IebVo9SaepRaMx2Jh2rhwV4+TpDKZSAATDairR0iiAMIQNpakBwLgAViWjQeABVyakdfjTs0AJJQHqwcpgCi0MIAQUT5Rg2TRuIeWi0QA)
```sh
--parser babel
```

**Input:**
```jsx
a = 
  {// prettier-ignore-start
  a,
// prettier-ignore-end
                                    };

a = // prettier-ignore-start
  {
  a,
// prettier-ignore-end
                                     };
```

**Output:**
```jsx
a = {
  // prettier-ignore-start
  a,
  // prettier-ignore-end
};

a = // prettier-ignore-start
  {
  a,
// prettier-ignore-end
                                     };  // <- This

```

chaitan94 commented:
It seems this test case doesn't work either:

[Playground Link](https://deploy-preview-8926--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEBDABAXne4AdKHAeiPQAcAnOGGASzgoFpaBzKCKxgZxlQpgI5UAGnQAjUTinSwogCbDB6EuSo16TVu04I5BAL4BuAgQxYAjIZDCQEMnWhdkoPhQgB3AAp8ETlKgAbd1QATycbMQpUMABragBlMmjaKBZkGAoAVzgbOABbMTg5OSKAGVRUzNQWOAAxDjzUdVTkNEyYCGsQAAsYPICAdW7aeC4ksDh43xHaADcRkNawLnCQFK4GGE8olkbkADNAjZsAKy4ADwAhKNiE1Dy4UpS4A6OckDPz+JSWALgARUyEHgrwCxxASQoGworTEqEKAS6lBSMAGtDkMG6yAAHAAGGyUCAbAZRMitShwaGzF42ACOQPg2zsfjQXEYUDgRSKXSo9NoVG21T2SEOYPeGzytHSWXFPz+gOBLxFbxsvDEaIxWKQACZVVFaAEfgBhCB5YUgSkAVi6mQ2ABV4X5ReDZtkAJJQEqweJgCi0ewAQU98RgIT+oI2+n0QA)

### Input
```js
a =  {
  // prettier-ignore-start
  a, b,          c, d,
  // prettier-ignore-end
};

a  =1;
```
### Output
```js
a = {
  // prettier-ignore-start
  a,
  b,
  c,
  d,
  // prettier-ignore-end
};

a  =1;
```

I think it's definitely worth fixing this if it were straightforward. I tried fixing this, maybe I am missing something here, but from what I can see, it does not really look like this can be done with simple changes.

In my opinion, I feel that range ignore would be generally used on large blocks of code on the outer scope. I feel these use cases are more appropriate for plain `// prettier-ignore`. Considering that, do you still think this is a realistic test case which we should be testing against?

evilebottnawi commented:
Ideally we should support it everywhere, so it's very good that we catch them now, if it is difficult to fix them we can open a new issue on these cases

chaitan94 commented:
Sure, I think that is a good idea. That way, we could at least get this existing functionality in place, and then continue to work on ironing out these cases in a new issue/PR.

evilebottnawi commented:
@chaitan94 Can we fix case above in this PR?

chaitan94 commented:
At the moment, from what I can see, I don't think there is a straightforward fix for the above case. But then again, I am new to the codebase so I can't say this for sure.

For more detail: The approach I am trying is to remember the indent size when we encounter `prettier-ignore-start`, and use that to restore it back when encountering `prettier-ignore-end`. I am able to do the first part (detecting and storing indent size), but I am not sure how to restore alignment from within `callPluginPrintFunction`. I am trying to use `addAlignmentToDoc`, but it doesn't seem to help.

evilebottnawi commented:
/cc @fisker What do you think should we merge as is and improve it in the future, or fix case here?

fisker commented:
This is quit a big  change, I'll need more time to test.

fisker commented on src/language-js/utils.js:
```diff
@@ -657,7 +657,8 @@ function hasJsxIgnoreComment(path) {
       (comment) => comment.value.trim() === "prettier-ignore"
-    )
+    ) &&
+    "next"
```
??

chaitan94 replied:
This is to return `"next"` instead of `true`.

fisker replied:
Sorry, I thought it's a `if`, but better rewrite to `if/else`

```js
return (foo && "next")
```

seems wired to me.

chaitan94 replied:
Sure, should I do `return (foo ? "next" : false)` instead, if that is more readable?

fisker replied:
```js
  if (prevSibling &&
    prevSibling.type === "JSXExpressionContainer" &&
    prevSibling.expression.type === "JSXEmptyExpression" &&
    prevSibling.expression.comments &&
    prevSibling.expression.comments.some(
      (comment) => comment.value.trim() === "prettier-ignore"
    )) {
    return "next";
  }

  return false;
```

chaitan94 replied:
Done.

evilebottnawi commented:
@fisker will you take care of the tests? no rush, just questions?

fisker commented:
I'll see what I can do, when I got time.

fisker commented:
**Prettier pr-8926**
[Playground link](https://deploy-preview-8926--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAzArlMMCW0AEqEEAFMADpQD0V+ADgE5wy5wMC0OA5lBE+wGcYAQwYxKAIwC8+AIyUa9JixxtOPPnHYIAJpXwHDRwwF8AlPgpQTlShiy4CRUlYnHjM2fkUB3ABYAnvgwfjgC+LwwhHwAtsLwelDuyeaWlCYgADQgEHSOUALIoKIMED4ACqIIhSjCADY+wgGF2RIMwmAA1swAynQdOFBcyDAM6HDZcDEScDo6swAywkPowlxwAGKx8bhDyCDC6DAQWSB+MDF1AOqh8AL9YHA91Ti4AG6vAftgAi0ggwI2DByu0uHFkKh6oDsgArAQADwAQu0ur1hDE4AtBnAIVCJiA4fCeoMuHU4ABFdAQeC4urQkD9BiAhj7CTCGZ1U6MQYwK44HQhZAADgADNlGBBAVd2nR9ow4My3jjsgBHKnwEG5GoHATsKBwWazU5MNU4JggtbgpCQun4wExHAjMZ2klkynUnHWvHZEQSPkCvzIABMPvaODqJIAwhAYlaQAqAKyndCAgAq7JqNvpb3GAEkoPNYD0wAwcHkAIIFnowAJk2mAkwmIA)
```sh
--parser babel
```

**Input:**
```jsx
function foo({
// prettier-ignore-start
b= 1
// prettier-ignore-end
             }) {
}

function foo({
b          = 1 // why this not formated
             }) {
}
```

**Output:**
```jsx
function foo({
  // prettier-ignore-start
  b= 1,
  // prettier-ignore-end
}) {
}

function foo({
b          = 1 // why this not formated
             }) {
}

```

fisker commented:
If I understand correctly , current implementation rely on attached comments, I don't think this not the right way, we move comments around sometimes.

I suggest calculate the ignore ranges from all comments, and check if node is inside this ranges.
Just an idea, not sure how will it go.

averydev commented:
Thanks for working on this @evilebottnawi this is the big reason Prettier isn't implemented in our codebase...

evilebottnawi commented:
@averydev Can we try to implement https://github.com/prettier/prettier/pull/8926#issuecomment-674356739 suggestion?

reduardo7 commented:
Check https://prettier.io/docs/en/ignore.html

gasparandr commented:
Thanks for giving this a shot.

rbalicki2 commented:
Has there been any progress on this issue? Is there currently any way to indicate that a JavaScript file is generated and should not be formatted in the file itself? I do not want to force users to add paths to their .prettierignore.
